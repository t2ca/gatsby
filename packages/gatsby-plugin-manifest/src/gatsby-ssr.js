import React from "react"
import { withPrefix } from "gatsby"
import { defaultIcons, createContentDigest, addDigestToPath } from "./common.js"
import fs from "fs"

let iconDigest = null

exports.onRenderBody = ({ setHeadComponents }, pluginOptions) => {
  // We use this to build a final array to pass as the argument to setHeadComponents at the end of onRenderBody.
  let headComponents = []

  const icons = pluginOptions.icons || defaultIcons
  const legacy =
    typeof pluginOptions.legacy !== `undefined` ? pluginOptions.legacy : true

  const cacheBusting =
    typeof pluginOptions.cache_busting_mode !== `undefined`
      ? pluginOptions.cache_busting_mode
      : `query`

  // If icons were generated, also add a favicon link.
  if (pluginOptions.icon) {
    let favicon = icons && icons.length ? icons[0].src : null

    if (cacheBusting !== `none`) {
      iconDigest = createContentDigest(fs.readFileSync(pluginOptions.icon))
    }

    const insertFaviconLinkTag =
      typeof pluginOptions.include_favicon !== `undefined`
        ? pluginOptions.include_favicon
        : true

    if (favicon && insertFaviconLinkTag) {
      headComponents.push(
        <link
          key={`gatsby-plugin-manifest-icon-link`}
          rel="shortcut icon"
          href={withPrefix(addDigestToPath(favicon, iconDigest, cacheBusting))}
        />
      )
    }
  }

  // Add manifest link tag.
  headComponents.push(
    <link
      key={`gatsby-plugin-manifest-link`}
      rel="manifest"
      href={withPrefix(`/manifest.webmanifest`)}
      crossOrigin={pluginOptions.crossOrigin}
    />
  )

  // The user has an option to opt out of the theme_color meta tag being inserted into the head.
  if (pluginOptions.theme_color) {
    let insertMetaTag =
      typeof pluginOptions.theme_color_in_head !== `undefined`
        ? pluginOptions.theme_color_in_head
        : true

    if (insertMetaTag) {
      headComponents.push(
        <meta
          key={`gatsby-plugin-manifest-meta`}
          name="theme-color"
          content={pluginOptions.theme_color}
        />
      )
    }
  }

  if (legacy) {
    const iconLinkTags = icons.map(icon => (
      <link
        key={`gatsby-plugin-manifest-apple-touch-icon-${icon.sizes}`}
        rel="apple-touch-icon"
        sizes={icon.sizes}
        href={withPrefix(addDigestToPath(icon.src, iconDigest, cacheBusting))}
      />
    ))

    headComponents = [...headComponents, ...iconLinkTags]
  }

  setHeadComponents(headComponents)
}
