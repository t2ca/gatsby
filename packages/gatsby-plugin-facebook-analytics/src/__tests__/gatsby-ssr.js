jest.mock(`common-tags`, () => {
  return {
    stripIndent: args => args,
  }
})

import { onRenderBody } from "../gatsby-ssr"

describe(`gatsby-plugin-facebook-analytics`, () => {
  describe(`onRenderBody`, () => {
    describe(`in development mode`, () => {
      it(`does not set any post body components`, () => {
        const setPostBodyComponents = jest.fn()

        onRenderBody({ setPostBodyComponents }, {})

        expect(setPostBodyComponents).not.toHaveBeenCalled()
      })
    })

    describe(`in production mode`, () => {
      let env

      beforeAll(() => {
        env = process.env.NODE_ENV
        process.env.NODE_ENV = `production`
      })

      afterAll(() => {
        process.env.NODE_ENV = env
      })

      it(`sets the correct post body components`, () => {
        const setPostBodyComponents = jest.fn()
        const pluginOptions = { appId: 1 }

        onRenderBody({ setPostBodyComponents }, pluginOptions)

        expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
      })

      it(`sets the correct script src during debug`, () => {
        const setPostBodyComponents = jest.fn()
        const pluginOptions = { appId: 1, debug: true }

        onRenderBody({ setPostBodyComponents }, pluginOptions)

        expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
      })

      it(`sets the correct script src for other languages than en_US`, () => {
        const setPostBodyComponents = jest.fn()
        const pluginOptions = { appId: 1, language: `sv` }

        onRenderBody({ setPostBodyComponents }, pluginOptions)

        expect(setPostBodyComponents.mock.calls).toMatchSnapshot()
      })
    })
  })
})
