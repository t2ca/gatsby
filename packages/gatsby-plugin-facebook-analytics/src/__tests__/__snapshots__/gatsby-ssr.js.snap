// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-facebook-analytics onRenderBody in production mode sets the correct post body components 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": Array [
              "
          window.fbAsyncInit = function() {
            FB.init({
              appId      : ",
              ",
              xfbml      : true,
              version    : 'v2.12'
            });

            FB.AppEvents.logPageView();

          };

          (function(d, s, id){
             var js, fjs = d.getElementsByTagName(s)[0];
             if (d.getElementById(id)) {return;}
             js = d.createElement(s); js.id = id;
             js.src = \\"https://connect.facebook.net/",
              "/sdk",
              "\\";
             fjs.parentNode.insertBefore(js, fjs);
           }(document, 'script', 'facebook-jssdk'));",
            ],
          }
        }
      />,
    ],
  ],
]
`;

exports[`gatsby-plugin-facebook-analytics onRenderBody in production mode sets the correct script src during debug 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": Array [
              "
          window.fbAsyncInit = function() {
            FB.init({
              appId      : ",
              ",
              xfbml      : true,
              version    : 'v2.12'
            });

            FB.AppEvents.logPageView();

          };

          (function(d, s, id){
             var js, fjs = d.getElementsByTagName(s)[0];
             if (d.getElementById(id)) {return;}
             js = d.createElement(s); js.id = id;
             js.src = \\"https://connect.facebook.net/",
              "/sdk",
              "\\";
             fjs.parentNode.insertBefore(js, fjs);
           }(document, 'script', 'facebook-jssdk'));",
            ],
          }
        }
      />,
    ],
  ],
]
`;

exports[`gatsby-plugin-facebook-analytics onRenderBody in production mode sets the correct script src for other languages than en_US 1`] = `
Array [
  Array [
    Array [
      <script
        dangerouslySetInnerHTML={
          Object {
            "__html": Array [
              "
          window.fbAsyncInit = function() {
            FB.init({
              appId      : ",
              ",
              xfbml      : true,
              version    : 'v2.12'
            });

            FB.AppEvents.logPageView();

          };

          (function(d, s, id){
             var js, fjs = d.getElementsByTagName(s)[0];
             if (d.getElementById(id)) {return;}
             js = d.createElement(s); js.id = id;
             js.src = \\"https://connect.facebook.net/",
              "/sdk",
              "\\";
             fjs.parentNode.insertBefore(js, fjs);
           }(document, 'script', 'facebook-jssdk'));",
            ],
          }
        }
      />,
    ],
  ],
]
`;
